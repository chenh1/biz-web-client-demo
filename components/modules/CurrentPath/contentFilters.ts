import type { Archetypes } from "../../constants/archetypes";
import type { VideoType } from "../../../graphql/queries/videos";

interface ArchetypeAttributes {
  risk: number; 
  experience: number; 
  patience: number; 
  willingness: number;
}

export type UserProfile = ArchetypeAttributes & {
  archetype: Archetypes | undefined;
}


const filterByTagConditions = (videos: { attributes: VideoType }[], condition: (any) => boolean) => {
  return videos?.filter(({ attributes }) => {
    const { tags: tagsData } = attributes
    const attributesSet = tagsData?.data
    const tags = attributesSet?.map(({ attributes: { value } }) => value)
    return condition(tags)
  })
}

export const filterVideos = (videos: { attributes: VideoType }[], userProfile: UserProfile): { attributes: VideoType }[] => {
  const { archetype } = userProfile
  let filtered = [...videos]

  switch (archetype) {
    // only novice  
    case "Essentialist":
      filtered = filterByTagConditions(videos, tags => tags.includes('novice'))
    //All intro only, risk averse and risk moderate, all series
    case "Researcher":
      filtered = filterByTagConditions(videos, tags => tags.includes('introductory') && (tags.includes('risk_averse') || tags.includes('risk_moderate')))
    //All intro and details, risk averse and risk moderate, all series
    case "Explorer":
      filtered = filterByTagConditions(videos, tags => (tags.includes('introductory') || tags.includes('details')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')))
    //All details and additional, series 2-5
    case "Pragmatist":
      filtered = filterByTagConditions(videos, tags => (tags.includes('details') || tags.includes('additional')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')) && !tags.includes("novice"))
    //All details and additional, risk averse and risk moderate, series 4 & 5
    case "Veteran":
      filtered = filterByTagConditions(videos, tags => (tags.includes('details') || tags.includes('additional')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')) && !tags.includes("novice") && !tags.includes("beginner") && !tags.includes("intermediate"))
    //All details and additional, series 4 & 5
    case "Adventurer":
      filtered = filterByTagConditions(videos, tags => (tags.includes('details') || tags.includes('additional')) && !tags.includes("novice") && !tags.includes("beginner") && !tags.includes("intermediate"))
    //Series 3-5
    case "Enthusiast":
      filtered = filterByTagConditions(videos, tags => !tags.includes("novice") && !tags.includes("beginner"))
    //All intro and details and additional, risk averse and risk moderate, all series
    case "Trailblazer":
      filtered = filterByTagConditions(videos, tags => (tags.includes('introductory') || tags.includes('details') || tags.includes('additional')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')))
    //All videos
    case "Scholar":
      filtered = [...videos]
  }

  return filtered;
}

export const getArchetype = ({ risk, experience, patience, willingness }: ArchetypeAttributes): Archetypes => {
  // only novice
  if (risk === 1 && experience === 1 && (patience <= 3 || willingness <= 3)) {
    return "Essentialist"
  }

  //All intro only, risk averse and risk moderate, all series
  if (risk <= 2 && experience === 1 && (patience <= 4 || willingness <= 4)) {
    return "Researcher"
  }

  //All intro and details, risk averse and risk moderate, all series
  if (risk <= 2 && experience === 1 && (patience <= 5 || willingness <= 5)) {
    return "Explorer"
  }

  //All details and additional, series 2-5
  if (
    (experience === 2 && (patience <= 3 || willingness <= 3)) ||
    (experience === 3 && (patience <= 4 || willingness <= 4))
  ) {
    return "Pragmatist"
  }

  //All details and additional, risk averse and risk moderate, series 4 & 5
  if (risk <= 2 && experience === 4 && (patience <= 4 || willingness <= 4)) {
    return "Veteran"
  }

  //All details and additional, series 4 & 5
  if (risk === 3 && experience === 4 && (patience <= 4 || willingness <= 4)) {
    return "Adventurer"
  }

  //Series 3-5
  if (experience === 4 && (patience > 4 || willingness > 4)) {
    return "Enthusiast"
  }

  //All intro and details and additional, risk averse and risk moderate, all series
  if (risk === 1 && experience === 1 && (patience > 5 || willingness > 5)) {
    return "Trailblazer"
  }

  //All videos
  if (risk >= 2 && experience <= 2 && (patience > 5 || willingness > 5)) {
    return "Scholar"
  }

  return "Scholar"
}