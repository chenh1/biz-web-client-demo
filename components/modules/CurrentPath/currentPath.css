@keyframes scroll-wheel {
  0% {
    transform: translateY(0) scaleY(1);
    opacity: 1;
  }
  100% {
    transform: translateY(-100%) scaleY(0.2);
    opacity: 0;
  }
}


.scroll-wheel-graphic {
  height: 16px;
  border: 4px solid #F3F4F6;
  border-radius: 50px;
  width: 1px;
  animation: scroll-wheel 1s linear infinite;
}