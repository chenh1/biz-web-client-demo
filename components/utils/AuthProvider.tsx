'use client'

import React from 'react';
import { SessionProvider } from 'next-auth/react'

const AuthProvider: React.FC<{ children?: string | JSX.Element | JSX.Element[] | React.ReactNode }> = ({ children }) => {
    return (
        <SessionProvider>
            {children}
        </SessionProvider>
    )
}

export default AuthProvider