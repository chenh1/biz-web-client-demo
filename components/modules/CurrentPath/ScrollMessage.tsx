'use client'

import React, { useState, useEffect } from "react";
import { Box } from "../../containers/Box/Box";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import './currentPath.css';
import { useIntersectionObserver } from "../../hooks/useIntersectionObserver";

export const ScrollMessage: React.FC = () => {
  const [ showScrollDownMessage, setShowScrollDownMessage ] = useState<boolean>(false)
  const [ isMobile, setIsMobile ] = useState<boolean>(false)
  
  useEffect(() => {
    // scroll listener with scroll top is x number of pixels equal to 20% of the viewport height
    if (typeof window !== 'undefined') {
      if (!showScrollDownMessage) {
        setTimeout(() => {
          if (window.scrollY <= window.innerHeight * .2) {
            setShowScrollDownMessage(true)
          }
        }, 5000)
      }

      setIsMobile(window.innerWidth < 640)
    }
  }, [])

  const moduleRef = useIntersectionObserver({
    inViewCallback: () => {
      setShowScrollDownMessage(false)
    }
  })

  return (
    <Box padding="0" gap="0" className={`absolute translate-y-full bottom-4 left-1/4 transition-opacity duration-[3000ms] ${showScrollDownMessage ? 'opacity-100' : 'opacity-0'}`}>
      <Box padding="0" gap="xs" className="relative place-items-center">
        <div className="scroll-wheel-graphic"></div>
        <Paragraph size="sm" className="italic">
          {isMobile ? "Swipe up to see all modules" : "Scroll down to see all modules"}
        </Paragraph>
        {/* scroll marker */}
        <div ref={moduleRef} className="absolute top-[300%]"></div>
      </Box>
    </Box>
  )
}
