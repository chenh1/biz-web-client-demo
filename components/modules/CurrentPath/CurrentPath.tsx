'use client'

import type { VideoType, Video } from "../../../graphql/queries/videos";
import type { SessionWithJwt } from "@/components/constants/types";
import type { RawTree } from "../../../graphql/queries/learningPaths";
import React, { useState, useEffect } from "react";
import { useLazyQuery, useQuery, useMutation } from "@apollo/client";
import pathQuestionnaireQuery from "../../../graphql/queries/pathQuestionnaire";
import { useSession } from "next-auth/react";
import { useRouter, useSearchParams } from "next/navigation";
import { myDataQuery } from "../../../graphql/queries/myData";
import { videosQuery } from "../../../graphql/queries/videos";
import { buildLinearTree } from "../../utils/treeDataUtils";
import { Heading } from "../../core/Heading/Heading";
import { Box } from "../../containers/Box/Box";
import colors from "../../constants/colors";
import { Paragraph } from "../../core/Paragraph/Paragraph";
import { Button } from "../../core/Button/Button";
import getScoreRank from "../../utils/getScoreRank";
import { updateMeMutation } from "../../../graphql/mutations/updateMe";
import { Loader } from "../../modules/Loader/Loader";
import { Overlay } from "../../containers/Overlay/Overlay";
import { ListProgress } from "../ListProgress/ListProgress";
import { allLearningPathsQuery } from "../../../graphql/queries/learningPaths";
import { ScrollMessage } from "./ScrollMessage";
import { archetypeIcons, archetypeDescriptions } from "../../constants/archetypes";
import { LoginForm } from "../LoginForm/LoginForm";
import './currentPath.css';

type Archetypes =
  "Essentialist" | 
  "Researcher" | 
  "Explorer" | 
  "Pragmatist" | 
  "Veteran" | 
  "Adventurer" | 
  "Enthusiast" | 
  "Trailblazer" | 
  "Scholar"

interface ArchetypeAttributes {
  risk: number; 
  experience: number; 
  patience: number; 
  willingness: number;
}

type UserProfile = ArchetypeAttributes & {
  archetype: Archetypes | undefined;
}


const filterByTagConditions = (videos: { attributes: VideoType }[], condition: (any) => boolean) => {
  return videos?.filter(({ attributes }) => {
    const { tags: tagsData } = attributes
    const attributesSet = tagsData?.data
    const tags = attributesSet?.map(({ attributes: { value } }) => value)
    return condition(tags)
  })
}

const filterVideos = (videos: { attributes: VideoType }[], userProfile: UserProfile): { attributes: VideoType }[] => {
  const { archetype } = userProfile
  let filtered = [...videos]

  switch (archetype) {
    // only novice  
    case "Essentialist":
      filtered = filterByTagConditions(videos, tags => tags.includes('novice'))
    //All intro only, risk averse and risk moderate, all series
    case "Researcher":
      filtered = filterByTagConditions(videos, tags => tags.includes('introductory') && (tags.includes('risk_averse') || tags.includes('risk_moderate')))
    //All intro and details, risk averse and risk moderate, all series
    case "Explorer":
      filtered = filterByTagConditions(videos, tags => (tags.includes('introductory') || tags.includes('details')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')))
    //All details and additional, series 2-5
    case "Pragmatist":
      filtered = filterByTagConditions(videos, tags => (tags.includes('details') || tags.includes('additional')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')) && !tags.includes("novice"))
    //All details and additional, risk averse and risk moderate, series 4 & 5
    case "Veteran":
      filtered = filterByTagConditions(videos, tags => (tags.includes('details') || tags.includes('additional')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')) && !tags.includes("novice") && !tags.includes("beginner") && !tags.includes("intermediate"))
    //All details and additional, series 4 & 5
    case "Adventurer":
      filtered = filterByTagConditions(videos, tags => (tags.includes('details') || tags.includes('additional')) && !tags.includes("novice") && !tags.includes("beginner") && !tags.includes("intermediate"))
    //Series 3-5
    case "Enthusiast":
      filtered = filterByTagConditions(videos, tags => !tags.includes("novice") && !tags.includes("beginner"))
    //All intro and details and additional, risk averse and risk moderate, all series
    case "Trailblazer":
      filtered = filterByTagConditions(videos, tags => (tags.includes('introductory') || tags.includes('details') || tags.includes('additional')) && (tags.includes('risk_averse') || tags.includes('risk_moderate')))
    //All videos
    case "Scholar":
      filtered = [...videos]
  }

  return filtered;
}

export const getArchetype = ({ risk, experience, patience, willingness }: ArchetypeAttributes): Archetypes => {
  // only novice
  if (risk === 1 && experience === 1 && (patience <= 3 || willingness <= 3)) {
    return "Essentialist"
  }

  //All intro only, risk averse and risk moderate, all series
  if (risk <= 2 && experience === 1 && (patience <= 4 || willingness <= 4)) {
    return "Researcher"
  }

  //All intro and details, risk averse and risk moderate, all series
  if (risk <= 2 && experience === 1 && (patience <= 5 || willingness <= 5)) {
    return "Explorer"
  }

  //All details and additional, series 2-5
  if (
    (experience === 2 && (patience <= 3 || willingness <= 3)) ||
    (experience === 3 && (patience <= 4 || willingness <= 4))
  ) {
    return "Pragmatist"
  }

  //All details and additional, risk averse and risk moderate, series 4 & 5
  if (risk <= 2 && experience === 4 && (patience <= 4 || willingness <= 4)) {
    return "Veteran"
  }

  //All details and additional, series 4 & 5
  if (risk === 3 && experience === 4 && (patience <= 4 || willingness <= 4)) {
    return "Adventurer"
  }

  //Series 3-5
  if (experience === 4 && (patience > 4 || willingness > 4)) {
    return "Enthusiast"
  }

  //All intro and details and additional, risk averse and risk moderate, all series
  if (risk === 1 && experience === 1 && (patience > 5 || willingness > 5)) {
    return "Trailblazer"
  }

  //All videos
  if (risk >= 2 && experience <= 2 && (patience > 5 || willingness > 5)) {
    return "Scholar"
  }

  return "Scholar"
}

export const CurrentPath = () => {
  const session = useSession().data as SessionWithJwt | null;
  const searchParams = useSearchParams()
  const router = useRouter()
  const searchParamKeys = [...(searchParams?.keys() || [])]
  const [ showScrollDownMessage, setShowScrollDownMessage ] = useState<boolean>(false)
  const [ showFullyCompletedMessage, setShowFullyCompletedMessage ] = useState<boolean>(searchParams?.get('allModulesCompleted') === 'true')
  const [ showPersonalPathCompletedMessage, setShowPersonalPathCompletedMessage ] = useState<boolean>(searchParams?.get('customPathCompleted') === 'true')
  const [ showArchetypeOverlay, setShowArchetypeOverlay ] = useState<boolean>(false)
  const [ constructedPath, setConstructedPath ] = useState<string | null | RawTree>(null)
  const [ pathQuestionnaireLoaded, setPathQuestionnaireLoaded ] = useState<boolean>(false)
  const [ nodeCount, setNodeCount ] = useState<number>(0)
  const [ nodeData, setNodeData ] = useState<{ data: RawTree } | null>(null)
  const [ isLoggingIn, setIsLoggingIn ] = useState<boolean>(false)
  const [ isMemberButNeverDidQuestionnaire, setIsMemberButNeverDidQuestionnaire ] = useState<boolean>(false)
  const [ userProfile, setUserProfile ] = useState<UserProfile>({
    risk: 0,
    willingness: 0,
    patience: 0,
    experience: 0,
    archetype: undefined
  })

  const [ updateMe ] = useMutation(updateMeMutation, {
    context: {
      headers: {
        'Authorization': `Bearer ${session?.data?.jwt}`
      }
    }
  })
  const { data: { videos: { data: videos = [] } = {} } = {} } = useQuery(videosQuery, { variables: { firstModule: 0, lastModule: 999 } })
  const [loadPathQuestionnaire, { data: { pathQuestionnaire: { data: { attributes: pathQuestionnaire = {} } = {} } = {} } = {} }] = useLazyQuery(pathQuestionnaireQuery);
  const { data: { learningPaths: { data: allLearningPaths = {} } = {} } = {} } = useQuery(allLearningPathsQuery)

  const { data: { myData: { progressData = {} } = {} } = {} } = useQuery(myDataQuery, {
    context: {
      headers: {
        'Authorization': `Bearer ${session?.data?.jwt}`
      },
    },
    skip: session?.status !== 'authenticated',
    fetchPolicy: 'network-only'
  })

  //Once pathQuestionnaire is loaded, build user profile and filter videos
  useEffect(() => {
    if (!!pathQuestionnaire) {
      const { pathQuestionnaireMappings } = pathQuestionnaire
      const newUserAttributes = {...userProfile}

      searchParamKeys.forEach(key => {
        if (key.length === 2 && key[0] === 'q') {
          const answerKey = searchParams?.get(key)
          const questionNumber = parseInt(key.replace('q', ''))
          const targetQuestion = pathQuestionnaireMappings.find(({ questionNumber: qn }) => qn === questionNumber)
          const attributes = targetQuestion?.answerToAttribute?.filter(({ answerKey: key }) => key === answerKey)

          attributes?.forEach(({ attribute, points }) => {
            newUserAttributes[attribute] += points
          })
        }
      })

      newUserAttributes.archetype = getArchetype(newUserAttributes)
      
      setShowArchetypeOverlay(true)
      setUserProfile(newUserAttributes)
    }
  }, [pathQuestionnaire])

  useEffect(() => {
    if (userProfile?.risk > 0 && userProfile?.willingness > 0 && userProfile?.patience > 0 && userProfile?.experience > 0 && !!videos) {
      const filteredVideos = filterVideos(videos, userProfile)
      //const filteredVideos = videos
      const rawTree = buildLinearTree(filteredVideos, progressData)
      setNodeCount(filteredVideos?.length)
      setConstructedPath(rawTree)
    }
  }, [ videos, userProfile, setUserProfile, pathQuestionnaireLoaded, setPathQuestionnaireLoaded ])

  useEffect(() => {
    if (!showScrollDownMessage) {
      setTimeout(() => {
        if (window.scrollY <= window.innerHeight * .2) {
          setShowScrollDownMessage(true)
        }
      }, 5000)
    }
  }, [])


  // if navigating from questionnaire (new user), process questionnaire and set path
  useEffect(() => {
    // this is a user who has just completed the questionnaire for the first time
    if (session?.status !== 'authenticated' && searchParamKeys?.length > 0 && !pathQuestionnaireLoaded) {
      // check if searchParamKeys contains any of the following: "q1", "q2", "q3", "q4"
      const hasQuestionnaireKeys = searchParamKeys?.some(key => key.length === 2 && key[0] === 'q')
      if (hasQuestionnaireKeys) {
        loadPathQuestionnaire()
        setPathQuestionnaireLoaded(true)
      }
    }

    // this could be a user who is already a member but is not logged in
    if (session?.status !== 'authenticated' && searchParamKeys?.length === 0) {
      //prompt login
    }
  }, [session?.status, searchParamKeys])

  // sync user progress to generated tree
  useEffect(() => {
    if (!progressData?.customPath && constructedPath && session?.status === 'authenticated') {
      updateMe({ variables: { progressData: { archetype: userProfile?.archetype, customPath: constructedPath } } })
    }
  }, [ progressData, constructedPath, session?.status ])

  
  useEffect(() => {
    // scroll listener with scroll top is x number of pixels equal to 20% of the viewport height
    const scrollListener = () => {
      if (window.scrollY > window.innerHeight * .2) {
        setShowScrollDownMessage(false)
      }
    }

    window.addEventListener('scroll', scrollListener)

    return () => {
      window.removeEventListener('scroll', scrollListener)
    }
  }, [])
  

  // prompt user to sign in if unauthenticated. User can choose not to sign in but let them know they won't be able to save progress
  const learningPathTree = progressData?.customPath || constructedPath

  const { attributes: previewModule } = videos?.filter(({ attributes: { lessonModule } }) => lessonModule == nodeData?.data.name)[0] || {}
  const { attributes: currentModule } = videos?.filter(({ attributes: { lessonModule } }) => lessonModule == progressData?.currentModule?.lessonModule)[0] || {}
  const examScore = progressData?.completedModules?.filter(({ lessonModule }) => lessonModule == nodeData?.data.name)[0]?.examScore
  const rank = getScoreRank(examScore)

  const nameMapCallback = (nameMap, nodeName) => {
    return nameMap?.filter(({ attributes: { lessonModule } }) => lessonModule?.toString() === nodeName?.toString())[0]?.attributes?.title
  }
  
  return (
    <Box padding="0" gap="xl" className="relative h-full sm:grid-cols-3">
      {isLoggingIn &&
        <Overlay wrapperClassName="px-4 w-full sm:w-auto" visibility={isLoggingIn} toggleVisibility={setIsLoggingIn}>
          <LoginForm defaultIsRegistering={false} toggleVisibility={setIsLoggingIn}/>
        </Overlay>
      }
      <Box padding="0" className="col-span-1 self-center">
        <Overlay visibility={showArchetypeOverlay} toggleVisibility={setShowArchetypeOverlay}>
          <Box padding="0" className="max-w-[640px]" gap="md">
            <Box padding="0" gap="md" className="sm:grid-cols-[160px,_auto] justify-center sm:justify-start items-center">
              <Box padding="0" className="justify-self-center">
                <img src={userProfile?.archetype ? archetypeIcons[userProfile?.archetype]: archetypeIcons["Scholar"]} alt="archetype icon" width={160} height={160}/>
              </Box>
              <Box padding="0" gap="sm" className="place-items-center sm:place-items-start">
                <Heading type="h4">Here are your results:</Heading>
                <Heading type="h2">{`You are the ${userProfile?.archetype}`}</Heading>
              </Box>
            </Box>
            <Heading type="h4">{userProfile?.archetype ? archetypeDescriptions[userProfile?.archetype]: archetypeDescriptions["Scholar"]}</Heading>
            <Paragraph className="italic">At this point, it's best to sign up now so you can save and track your progress. If you choose not to, it's ok. You can browse the modules but just keep in mind you won't be able to track your progress and your tailored path will be lost.</Paragraph>
            <Button onClick={() => {setShowArchetypeOverlay(false); setIsLoggingIn(true)}} label="Got it!" primary/>
          </Box>
        </Overlay>
        <Overlay visibility={showPersonalPathCompletedMessage} toggleVisibility={setShowPersonalPathCompletedMessage}>
          <Box className="max-w-[640px]">
            <Heading type="h2">You've completed your path! 🎓</Heading>
            <Heading type="h4">Thanks for letting us take you through this journey. It doesn't end here, though. Remember to practice, go through all of the module tests, and keep tuning back in. We'll provide more tools, tips, and regular content to keep you sharp and to continue guiding you through each week as you grow your portfolio! If you want to continue your learning beyond your path, scroll down to find more modules.</Heading>
            <Button onClick={() => setShowPersonalPathCompletedMessage(false)} label="Got it!" primary/>
          </Box>
        </Overlay>
        <Overlay visibility={showFullyCompletedMessage} toggleVisibility={setShowFullyCompletedMessage}>
          <Box className="max-w-[640px]">
            <Heading type="h2">You've completed all of the courses! 🎓</Heading>
            <Heading type="h4">What a long trip it's been. Thanks for letting us take you through this journey. It doesn't end here, though. Remember to practice, go through all of the module tests, and keep tuning back in. We'll provide more tools, tips, and regular content to keep you sharp and to continue guiding you through each week as you grow your portfolio!</Heading>
            <Button onClick={() => setShowFullyCompletedMessage(false)} label="Got it!" primary/>
          </Box>
        </Overlay>


        {session?.status !== 'authenticated' && session?.status !== 'loading' && !!learningPathTree &&
          <>
            <Heading type="h4" className="text-center sm:text-left" fontSizeClass="h4">
              Log in now to save and track your progress.
            </Heading>
            <Button data-cy="log-in-prompt-learning-path" label="Log in" primary onClick={() => setIsLoggingIn(true)}/>
          </>
        }
        <Heading data-cy="pick-up-left-off-text" className="text-center sm:text-left" type="h3">
          {!!currentModule ? 'Pick up where you left off' : 'Start your learning journey'}
        </Heading>
        {!!currentModule &&
          <>
            <Heading type="h4" className="text-center sm:text-left" >
              {currentModule?.title}
            </Heading>
            <Paragraph className="text-center sm:text-left" >
              {currentModule?.description}
            </Paragraph>
            <Button label="Start" primary onClick={() => router.push(`/module/${currentModule?.lessonModule?.toString().replace('.', '-')}`)}/>
          </>
        }
        {!!userProfile?.archetype &&
          <Box padding="0" className="place-self-center">
            <img src={archetypeIcons[userProfile?.archetype]} alt="archetype icon" width={160} height={160}/>
          </Box>
        }
      </Box>
      <Box padding="0" className="sm:col-span-2">
      {learningPathTree ?
        <ListProgress
          isVertical
          nameMapCallback={nameMapCallback}
          nameMap={videos}
          dataCy="current-path"
          globalOnMouseEnter={(node) => setNodeData(node)}
          progressData={progressData}
          //globalOnMouseLeave={() => setNodeData(null)}
          rawTree={learningPathTree}
          highlightValue={progressData?.currentModule?.lessonModule}
          className={`col-span-1 w-full min-h-[${nodeCount * 100}px]`}
          twFaderColor={"night-black"}
          background={colors.nightBlack}
          allLearningPaths={allLearningPaths}
          showAsList
          nodeWidth={12}
          nodeHeight={12} 
          availableModules={[]}        
        />
        :
        !currentModule ?
          <Box padding="0" className="place-items-center">
            <Box padding="0" gap="md" className="justify-center items-center">
              {session?.status === 'loading' && session?.data === null ?
                <Loader/>
                :
                <>
                  {session?.status !== 'authenticated' &&
                    <>
                      <Heading type="h4" className="text-center">Already have an account and you want to pick up where you left off?</Heading>
                      <Button size="lg" label="Yes, log me in" primary onClick={() => setIsLoggingIn(true)}/>
                      <Heading type="h4" className="text-center">If not, you can</Heading>
                    </>
                  }
                  <Button data-cy="tailor-lesson-button" size="lg" label="Tailor my lesson plan" primary={session?.status === 'authenticated'} onClick={() => router.push(`/path`)}/>
                  <Heading type="h4" className="text-center">or</Heading>
                  <Button data-cy="start-from-top-button" size="lg" label="Just start me off from the top" onClick={() => {
                    if (session?.status === 'authenticated') {
                      updateMe({ variables: { progressData: { customPath: null } } })
                    }
                    router.push(`/module/1-01`)
                  }}/>
                </>
              }
            </Box>
          </Box>
          :
          <></>
      }
      </Box>

      <ScrollMessage/>
    </Box>
  )
}
