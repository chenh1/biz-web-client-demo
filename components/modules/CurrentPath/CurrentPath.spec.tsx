import { CurrentPath } from "./CurrentPath";
import { MockedProvider } from "@apollo/client/testing";
import { render, waitFor } from '@testing-library/react';
import { JSDOM } from "jsdom"
import { videosQuery } from "../../../graphql/queries/videos";
import pathQuestionnaireQuery from "../../../graphql/queries/pathQuestionnaire";
import { allLearningPathsQuery } from "../../../graphql/queries/learningPaths";
import AuthProvider from "../../utils/AuthProvider";
import { myDataQuery } from "../../../graphql/queries/myData";
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window

const observe = jest.fn();
const unobserve = jest.fn();

// you can also pass the mock implementation
// to jest.fn as an argument
global.IntersectionObserver = jest.fn(function (callback, options) {
  return {
    observe,
    unobserve,
    root: null,
    rootMargin: '',
    thresholds: [],
    disconnect: jest.fn(),
    takeRecords: jest.fn(),
  };
});


jest.mock("next/navigation", () => ({
  useRouter() {
    return {
      prefetch: () => null
    };
  },
  useSearchParams: () => {
    return {
      get: () => "1",
      keys: () => []
    }
  }
}));

const mocks = [
  {
    request: {
      query: myDataQuery,
    },
    result: {
      data: {
        myData: {
          progressData: {
            currentModule: {
              lessonModule: 1.01
            },
            completedModules: [
              { lessonModule: 1.01, examScore: 100 },
              { lessonModule: 1.02, examScore: 100 }
            ]
          }
        }
      }
    }
  },
  {
    request: {
      query: allLearningPathsQuery
    },
    result: {
      data: {
        learningPaths: {
          data: [
            { attributes: { title: "Intro to cooking", level: "1", description: "Learn the basics of cooking", rawTree: JSON.parse(JSON.stringify({ name: "Lesson 1", description: "Some description", isCompleted: false, children: [ { name: "Lesson 2", description: "Some description", isCompleted: false, children: [] } ] })) } },
            { attributes: { title: "Intermediate cooking", level: "2", description: "Take the next steps", rawTree: JSON.parse(JSON.stringify({ name: "Lesson 1", description: "Some description", isCompleted: false, children: [ { name: "Lesson 2", description: "Some description", isCompleted: false, children: [] } ] })) } },
            { attributes: { title: "Master chef", level: "3", description: "Gain master in cooking", rawTree: JSON.parse(JSON.stringify({ name: "Lesson 1", description: "Some description", isCompleted: false, children: [ { name: "Lesson 2", description: "Some description", isCompleted: false, children: [] } ] })) } }
          ]
        }
      }
    }
  },
  {
    request: {
      query: pathQuestionnaireQuery
    },
    result: {
      data: {
        pathQuestionnaire: {
          data: [
            {
              attributes: {
                questions: [
                  {
                    question: "Question 1",
                    answers: [
                      {
                        key: "1",
                        answer: "Answer 1"
                      },
                      {
                        key: "2",
                        answer: "Answer 2"
                      }
                    ]
                  }
                ],
                pathQuestionnaireMappings: [
                  {
                    questionNumber: 1,
                    answerToAttribute: {
                      attribute: "attribute1",
                      points: 1,
                      answerKey: "1"
                    }
                  }
                ]
              }
            }
          ]
        }
      }
    }
  },
  {
    request: {
      query: videosQuery,
    },
    result: {
      data: {
        videos: {
          data: [
            {
              attributes: {
                title: "Title 1",
                url: "https://www.youtube.com/watch?v=1",
                description: "Description 1",
                transcript: "Transcript 1",
                lessonModule: 1.01,
                tags: {
                  data: [
                    {
                      attributes: {
                        displayName: "Tag 1",
                        color: "red",
                        value: "tag1"
                      }
                    }
                  ]
                }
              }
            },
            {
              attributes: {
                title: "Title 2",
                url: "https://www.youtube.com/watch?v=2",
                description: "Description 2",
                transcript: "Transcript 2",
                lessonModule: 1.02,
                tags: {
                  data: [
                    {
                      attributes: {
                        displayName: "Tag 2",
                        color: "blue",
                        value: "tag2"
                      }
                    }
                  ]
                }
              }
            }
          ]
        }
      }
    }
  }
];


it('renders multiple series of lessons in a vertical node format', async () => {
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <AuthProvider>
        <CurrentPath />
      </AuthProvider>
    </MockedProvider>
  );

  await waitFor(() => new Promise((res) => setTimeout(res, 800)));

  expect(container).toMatchSnapshot();
});
