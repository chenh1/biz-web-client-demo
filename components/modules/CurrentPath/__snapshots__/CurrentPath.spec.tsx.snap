// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders multiple series of lessons in a vertical node format 1`] = `
<div>
  <div
    class=" grid  gap-16  p-2 relative h-full sm:grid-cols-3"
  >
    <div
      class=" grid  gap-4  p-2 col-span-1 self-center"
    >
      <h3
        class="text-night-black dark:text-max-white font-headings font-semibold text-2xl sm:text-4xl text-center sm:text-left"
        data-cy="pick-up-left-off-text"
      >
        Start your learning journey
      </h3>
    </div>
    <div
      class=" grid  gap-4  p-2 sm:col-span-2"
    >
      <div
        class=" grid  gap-4  p-2 place-items-center"
      >
        <div
          class=" grid  gap-8  p-2 justify-center items-center"
        >
          <h4
            class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md text-center"
          >
            Already have an account and you want to pick up where you left off?
          </h4>
          <button
            class="   w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-max-white px-6 py-3   bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red"
            data-testid="button"
            type="button"
          >
            <div
              class="bg-hyper-red border-hyper-red dark:bg-hyper-red dark:border-hyper-red rounded-lg absolute left-0 top-0 w-full h-full "
            />
            <span
              class="pointer-events-none relative z-10"
            >
              Yes, log me in
            </span>
          </button>
          <h4
            class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md text-center"
          >
            If not, you can
          </h4>
          <button
            class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-6 py-3   "
            data-cy="tailor-lesson-button"
            data-testid="button"
            type="button"
          >
            <div
              class="triangle left"
            >
              <div />
            </div>
            <div
              class="triangle right"
            >
              <div />
            </div>
            <span
              class="pointer-events-none relative z-10"
            >
              Tailor my lesson plan
            </span>
          </button>
          <h4
            class="text-night-black dark:text-max-white font-headings font-semibold text-lg sm:text-md text-center"
          >
            or
          </h4>
          <button
            class=" secondary  w-full sm:w-auto cursor-pointer border-2 font-bold leading-none inline-block rounded-lg text-center font-headings relative text-slate-700 bg-transparent border-slate-700 dark:text-max-white dark:border-max-white active:text-max-white active:bg-slate-700 active:border-slate-700 dark:active:bg-max-white dark:active:border-max-white dark:active:text-night-black px-6 py-3   "
            data-cy="start-from-top-button"
            data-testid="button"
            type="button"
          >
            <div
              class="triangle left"
            >
              <div />
            </div>
            <div
              class="triangle right"
            >
              <div />
            </div>
            <span
              class="pointer-events-none relative z-10"
            >
              Just start me off from the top
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      class=" grid  gap-4  p-2 absolute translate-y-full bottom-4 left-1/4 transition-opacity duration-[3000ms] opacity-0"
    >
      <div
        class=" grid  gap-2  p-2 relative place-items-center"
      >
        <div
          class="scroll-wheel-graphic"
        />
        <p
          class=" text-night-black dark:text-max-white font-body font-body text-sm italic"
        >
          Scroll down to see all modules
        </p>
        <div
          class="absolute top-[300%]"
        />
      </div>
    </div>
  </div>
</div>
`;
